metadata-ingestion/src/datahub/ingestion/source/sql/mssql/source.py

```diff
@@ -0,0 +1,679 @@
+        return columns
+
+    def get_database_level_workunits(
+            self,
+            sql_config: SQLAlchemyConfig,
+            inspector: Inspector,
+            database: str,
+    ) -> Iterable[MetadataWorkUnit]:
+        yield from super().get_database_level_workunits(
+            sql_config=sql_config,
+            inspector=inspector,
+            database=database,
+        )
+        if sql_config.include_jobs:
+            yield from self.loop_jobs(inspector, sql_config)
+
+    def get_schema_level_workunits(
+            self,
+            sql_config: SQLAlchemyConfig,
+            inspector: Inspector,
+            schema: str,
+            database: str,
+    ) -> Iterable[Union[MetadataWorkUnit, SqlWorkUnit]]:
+        yield from super().get_schema_level_workunits(
+            sql_config=sql_config,
+            inspector=inspector,
+            schema=schema,
+            database=database,
+        )
+        if sql_config.include_stored_procedures:
+            yield from self.loop_stored_procedures(
+                inspector, schema, sql_config
+            )
+
+    def _get_jobs(self, conn: Connection, db_name: str) -> Dict[str, Dict[str, Any]]:
+        jobs_data = conn.execute(
+            f"""
+            SELECT
+                job.job_id,
+                job.name,
+                job.description,
+                job.date_created,
+                job.date_modified,
+                steps.step_id,
+                steps.step_name,
+                steps.subsystem,
+                steps.command,
+                steps.database_name
+            FROM
+                msdb.dbo.sysjobs job
+            INNER JOIN
+                msdb.dbo.sysjobsteps steps
+            ON
+                job.job_id = steps.job_id
+            where database_name = '{db_name}'
+            """
+        )
+        jobs: Dict[str, Dict[str, Any]] = {}
+        for row in jobs_data:
+            step_data = dict(
+                job_id=row["job_id"],
+                job_name=row["name"],
+                description=row["description"],
+                date_created=row["date_created"],
+                date_modified=row["date_modified"],
+                step_id=row["step_id"],
+                step_name=row["step_name"],
+                subsystem=row["subsystem"],
+                command=row["command"],
+            )
+            if row["name"] in jobs:
+                jobs[row["name"]][row["step_id"]] = step_data
+            else:
+                jobs[row["name"]] = {row["step_id"]: step_data}
+        return jobs
+
+    def loop_jobs(
+        self,
+        inspector: Inspector,
+        sql_config: SQLServerConfig,
+    ) -> Iterable[MetadataWorkUnit]:
+        """
+        Loop MS SQL jobs as dataFlow-s.
+        :return:
+        """
+        db_name = self.get_db_name(inspector)
+        with inspector.engine.connect() as conn:
+            jobs = self._get_jobs(conn, db_name)
```

Comment:
If the query inside `_get_jobs` fails, the ingestion will fail miserably without ingesting any table or view, that it could ingest before this PR.

Let's surround call to `loop_jobs` with try..except and report warnings as mentioned in my comment earlier https://github.com/datahub-project/datahub/pull/5363#discussion_r1071831016

Similarly call to `loop_stored_procedures` needs to be handled for exceptions. The intention is that -> even if the job or procedures ingestion fails due to query permission error or so, the rest of the ingestion (tables, views, etc) should continue as usual. Otherwise this PR will break ingestion for people who are using MSSQL source with credentials that do not have permissions to list jobs and stored procedures.

Another option is to keep jobs and stored procedures ingestion off by default in recipe and leaving it to users to enable this, if required. However even in those cases, it would be a lot better if ingestion does not fail completely but only partially in case of missing permissions.

Reply:
Updated

Reply:
Thank you.